src/vanilla/store.ts

```diff
@@ -482,13 +482,123 @@ const buildStore = (
     return [sorted, visited]
   }
 
+  function kahnsToposort(
+    pending: Pending,
+    rootAtom: AnyAtom,
+    rootAtomState: AtomState,
+  ): [[AnyAtom, AtomState, number][], Set<AnyAtom>] {
+    function createQueue<T>() {
+      let head = 0
+      let tail = 0
+      const items: T[] = []
+      const isEmpty = () => head === tail
+      const enqueue = (item: T) => {
+        items[tail++] = item
+      }
+      const dequeue = () => {
+        if (head === tail) {
+          return undefined
+        }
+        const item = items[head]
+        items[head++] = undefined as any
+        return item
+      }
+      return { enqueue, dequeue, isEmpty }
+    }
+    // Kahn's algorithm for topological sorting
+    const atomLookup = []
+    const stateMap: AtomState[] = [rootAtomState]
+    const adjacencyMap: number[][] = []
+    const visited = new Set([rootAtom])
+    {
+      // 1: build the dependency graph
+      // and calculate in-degrees (number of incomming edges aka dependencies)
+      let nextId = 0
+      const queue = createQueue<[a: AnyAtom, aState: AtomState, id: number]>()
+      queue.enqueue([rootAtom, rootAtomState, nextId])
+      while (!queue.isEmpty()) {
+        const [a, aState, id] = queue.dequeue()!
+        atomLookup[id] = a
+        stateMap[id] = aState
+        const dependentAtoms = (adjacencyMap[id] ??= [])
+        for (const [d, ds] of getDependents(pending, a, aState).entries()) {
+          if (visited.has(d)) {
+            continue
+          }
+          dependentAtoms.push(++nextId)
+          queue.enqueue([d, ds, nextId])
+          visited.add(d)
+        }
+      }
+    }
+    const inDegree: number[] = []
+    {
+      for (let i = 0; i < visited.size; i++) {
+        inDegree[i] = 0
+        for (const vertex of adjacencyMap[i]!) {
+          inDegree[vertex]!++
+        }
+      }
+    }
+    const queue = createQueue<number>()
+    {
+      // 2. Initialize queue with nodes that have no incoming edges
+      for (let i = 0; i < visited.size; i++) {
+        if (inDegree[i] === 0) {
+          queue.enqueue(i)
+        }
+      }
+    }
+    const sorted: [
+      atom: AnyAtom,
+      atomState: AtomState,
+      epochNumber: number, //
+    ][] = []
+    {
+      // 3. Process each node
+      while (!queue.isEmpty()) {
+        const a = queue.dequeue()!
+        const aState = stateMap[a]!
+        sorted.push([atomLookup[a]!, aState, aState.n])
+        for (const adjacent of adjacencyMap[a]!) {
+          inDegree[adjacent]!--
+          // If indegree becomes 0, push it to the queue
+          if (inDegree[adjacent] === 0) {
+            queue.enqueue(adjacent)
+          }
+        }
+      }
+    }
+    return [sorted, visited]
+  }
```

Comment:
@dai-shi Can I get your thoughts on this approach?

Should we keep the current iterative DFS approach, or do you think the performance gains are worth the added complexity of Kahns?

Kahns does marginally better on Star topologies, which I believe would be the common case.